box_error.go

```diff
@@ -0,0 +1,274 @@
+package tarantool
+
+import (
+	"fmt"
+)
+
+const errorExtID = 3
+
+// BoxError is a type representing Tarantool `box.error` object: a single
+// MP_ERROR_STACK object with a link to the previous stack error.
+// See https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_error/error/
+//
+// Since 1.10.0
+type BoxError struct {
+	// Type is error type that implies its source (for example, "ClientError").
+	Type string
+	// File is a source code file where the error was caught.
+	File string
+	// Line is a number of line in the source code file where the error was caught.
+	Line uint32
+	// Msg is the text of reason.
+	Msg string
+	// Errno is the ordinal number of the error.
+	Errno uint32
+	// Code is the number of the error as defined in `errcode.h`.
+	Code uint32
+	// Fields are additional fields depending on error type. For example, if
+	// type is "AccessDeniedError", then it will include "object_type",
+	// "object_name", "access_type".
+	Fields map[interface{}]interface{}
+	// Prev is the previous error in stack.
+	Prev *BoxError
+}
+
+// Error converts a BoxError to a string.
+func (boxError *BoxError) Error() string {
+	return fmt.Sprintf("%s (%s, code 0x%x), see %s:%d",
+		boxError.Msg, boxError.Type, boxError.Code,
+		boxError.File, boxError.Line)
```

Comment:
Added recursive print for `Prev` and unit test.

To be honest, I think `Fields` doesn't worth adding to. For example, the info stored in AccessDeniedError fields is a duplication of info already mentioned in the message. Like, it may be convenient if you want to inspect in code which exact user was rejected without parsing the message string, but it is rather useless in error printout. Same goes for constraint violation errors, see [field definitions](https://github.com/tarantool/tarantool/blob/c60a820506322419c9537c0cd641b13cc8999e0c/src/box/tuple_constraint_func.c#L104-L113) and [message example](https://github.com/tarantool/tarantool/blob/ed9b982d319ae4020e0536333c748526287a27f5/test/engine-luatest/gh_6436_field_constraint_test.lua#L96).